### DISCUSSION TITLE
Add job queue to reduce cloud costs

### DISCUSSION THREAD

**Discussion by u12206050** [directus/directus#19537] [Feature Requests]
### Summary

Due to the fantastic event and hook system in Directus it is unfortunately not entirely possible to only allocate cloud cpu during requests, which is often a setting that can be set between always being allocated vs only during requests.

Suggest an inbuilt solution to offload async tasks to the database or redis (very risky) and have them either be executed by a cron job every minute, or on trigger by PubSub.

### Motivation

The reason being that any hooks triggered asynchronously need the cpu in order to compute, and if there doesn't happen to be another request executing on the same container then those hooks never execute, or worse, they start but never get to finish.

This leaves us with the only option of always having the cpu allocated, which increases cloud costs by a factor of 30x since it is allocated the entire time the container instance is online.

### Detailed Design

Currently when an async event is triggered it could be said that it is delayed until the response is returned before getting executed, that means it is simply in memory until it is executed, and if it fails or succeeds is up to each developer to implement what should happen.

So, I suggest a persistent queue service that consists of a table, queue service and executor worker.

The queue table:

`id`: Can be a generated uuid from a hash of **action + item id(s)**
`action`: Event (eg. items_updated) or Custom Job (eg. sync_to_stripe)
`payload`: Event { collection, keys } or Custom Job (anything)
`status`: "pending | processing | failed | completed"
`output`: Error or success or just empty
`queued`: Number of times this job has been queued (Could be queued multiple times before it is processed once)
`processed`: Number of times this job has been processed
`failed`: Number of times it has failed
`queued_for` Current Time or future if job needs to be executed in future.
`last_failed`: Timestamp when it failed last
`last_completed`: Timestamp when it completed last
`date_created`
`date_updated`

The QueueService Interface. 

`enqueueEvent`: Happens on every action (create, update, delete) and is persisted to the queue table per item. Meaning even if an item is updated multiple times during a request, or until the jobs are executed, there will only be one item in the queue table, with status `pending`

`dequeueEvent`: Selects an event from the table and sets the status to `processing`. To reduce contention on the queue table, we can select a random `pending` job, if we have multiple workers executing jobs.

`enqueueJob`: For custom jobs/events that developers want to queue for a later time or for the next cycle. Each job has an action just like the generic items action, but is mapped back to a custom functions in the jobs directory. Status same as events

`dequeueJob`: Used by the jobs worker to dequeue and execute the job. Status same as events

`refreshQueue`: This will check any `processing` jobs in the queue and time them out if they have been running for too long.

Both custom jobs and events are executed by a queue worker, that can be executed using a cron job for example every minute. When a job/event is processed, it is executed inside a `try-catch` and on error the status is set to `failed`, and the error message is stored in the `output` column of the job/event

Optionally, if need be accountability could also be added.

### Must Have:
- Ability to run jobs in a separate request than the original request.
- No requirement for constantly having the cpu allocated

### Should Have:
- Interface to add custom jobs and possible events

### Could Have:
- Dashboard or panel to view job status and statistics

### Won't Have:
- -

### Drawbacks

Might cause a longer than expected delay in hook execution than the current implementation

### Alternatives

Alternatives is implementing offloading events directly to PubSub, such that PubSub can then call the service back with a request to execute any logic that might be defined.

Problem with this though is the complexity it adds to the entire architecture stack.

### Adoption Strategy

If implemented correctly, the only change would be for developers to set up an external(optional) cron job that makes a request for the job worker to start processing some jobs.

Events that are dequeued will then dispatch as they currently do, but instead of asynchronously it will be executed synchronously during the worker's request.

**Comment by rijkvanzanten** [comment]
> and have them either be executed by a cron job every minute, or on trigger by PubSub.

Wouldn't a pubsub trigger mean that the container will have to be running non-stop regardless in order to listen to the events? 🤔 

Separating out "workers" from the main server runtime would make sense here, where we'd effectively implement/rely on a queuing system for async operations

**Reply by u12206050** [reply]
Do note, that it doesn't have to be made all that complicated. As long as you can some how trigger your system with a request every minute the same as what a cron job would have done, you effectively will be able to solve this in the same system without a "separate" worker. There are a number of cron workers even some free ones that can be used to just make a request. The trick will be to not have the worker or request time out if the jobs are taking long.

**Reply by andreasgangso** [reply]
it would be very simple to setup on gcp with cloud run + cloud scheduler https://cloud.google.com/scheduler/docs/creating

**Comment by u12206050** [comment]
We would only subscribe a webhook to PubSub that it should request for each job, so then the container can scale to 0 and not need cpu until a new request is made.

**Comment by u12206050** [comment]
In terms of cron jobs, I currently use an external cloud schedular that calls a `jobs` endpoint every minute, which then executes up to a 100 tasks that are in the queue table. Unfortunately this isn't enough as events being emitted are still being emitted in async, and I could rewrite everything to just use hooks and not actions, but thought it might be worth looking at implementing this as optional in Directus Core

**Comment by Attacler** [comment]
Just an idea, maybe also add a job type and limits for that.
An example usecase would be that Stripe (and other services) have rate limits.
So if you can tell the job system that a certain action can only be invoked 100 times a minute, it would help in both distributing the load + take care of ratelimits that might occurre

**Comment by u12206050** [comment]
Good idea, although I wouldn't store that on each job to be executed, but rather as part of the executor config, since that is checked every minute, to see if it should actually check for jobs or not. Eg. some jobs could be cron based and only execute every 5minutes etc.

**Reply by Attacler** [reply]
I think for some cases it would be easier to do.
For example it could also be used to bundle requests to save load/execution time in the end.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by falmanna** [comment]
I am using bullmq in my bundle extension and so far it is doing very good. 

The only problem is when developing with extension reload, there is no obvious way to clean up workers as directus doesn't emit any events. So we are forced to restart the server every time we update a job. 

Is there a way to hook into the extension reload event to do some cleanups?

**Reply by rijkvanzanten** [reply]
> Is there a way to hook into the extension reload event to do some cleanups?

I don't think there's an existing event as of right now, but I'll happily take a PR to add one in. I'll spin up a new issue so we can keep discussion there 🙂  https://github.com/directus/directus/issues/24489

**Reply by falmanna** [reply]
@rijkvanzanten sounds cool. I will take a look

**Comment by u12206050** [comment]
I tested out a MVP with the filter hooks system of Directus. Basically, I subscribe to everything using the filter hook by Directus, that in turns publishes the event to pubsub with the payload, keys and accountability. Then in another part of the service I subscribe to the events I want (only needs to be done once every 30 days), the subscribe takes the collection name, action (create, delete, update) and the endpoint which should be called.

The endpoint then is called by pubsub when an event is received. This effectively means when a user makes and update (request 1), that update publishes to pubsub, which then checks for active subscriptions, it sees my subscription and calls it (request 2).
Request 1 (the user request) could already have been returned whilst
Request 2 (the pubsub request) could run as long as it needs to, and if it fails pubsub could try again. Also note this could also trigger other events again if needed.

Cost wise, I only pay for the duration of each of these requests then.

FYI, Redis could also be used for events, but I believe it will trigger all containers subscribed instead of just calling one.

**Reply by falmanna** [reply]
Interesting. I like the global action/filter dispatcher approach.

I actually integrated bullmq connected to the same redis instance. I created few utils for creating jobs and workers from directus actions. Basically, it wraps and action and gives another one that will run in a worker while being fully configurable. 
It is a really good library. It Supports crons, delayed jobs, retry, concurrency, dependency, workflows, and other nice things. Also for less demanding or local installations, it can run in the same process. 

Now, I am only having problems with local dev mode as there is no way (yet) to subscribe to extension reload event to do some cleanups. Will look at the issue above maybe I can push a PR.

**Reply by u12206050** [reply]
That sounds good, I'm going to check it out. Another alternative I quickly tried out was using dapr. This was much quicker to setup locally and has quite a few options as well.

**Reply by falmanna** [reply]
First time seeing dapr, looks similar to temporal. I opted for bullmq to keep things simple and in one instance.
